drivers/digital-io/max149x6/iio_max14916.c

```diff
@@ -0,0 +1,296 @@
+ */
+static int max14916_iio_read_raw(void *dev, char *buf, uint32_t len,
+				 const struct iio_ch_info *channel,
+				 intptr_t priv)
+{
+	struct max14916_iio_desc *desc = dev;
+	uint32_t val;
+	int ret;
+
+	ret = max14916_ch_get(desc->max14916_desc, channel->address, &val);
+	if (ret)
+		return ret;
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, (int32_t *)&val);
+}
+
+/**
+ * @brief Write the raw attribute for a specific channel
+ * @param dev - The iio device structure.
+ * @param buf - Buffer to be filled with requested data.
+ * @param len - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv - Private descriptor
+ * @return 0 in case of success, error code otherwise
+ */
+static int max14916_iio_write_raw(void *dev, char *buf, uint32_t len,
+				  const struct iio_ch_info *channel,
+				  intptr_t priv)
+{
+	struct max14916_iio_desc *desc = dev;
+	int32_t val;
+
+	iio_parse_value(buf, IIO_VAL_INT, &val, NULL);
+
+	return max14916_ch_set(desc->max14916_desc, channel->address, val);
+}
+
+/**
+ * @brief Read the offset attribute for a specific channel
+ * @param dev - The iio device structure.
+ * @param buf - Buffer to be filled with requested data.
+ * @param len - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv - Private descriptor
+ * @return 0 in case of success, error code otherwise
+ */
+static int max14916_iio_read_offset(void *dev, char *buf, uint32_t len,
+				    const struct iio_ch_info *channel,
+				    intptr_t priv)
+{
+	int32_t val = 0;
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+}
+
+/**
+ * @brief Read the scale attribute for a specific channel
+ * @param dev - The iio device structure.
+ * @param buf - Buffer to be filled with requested data.
+ * @param len - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv - Private descriptor
+ * @return 0 in case of success, error code otherwise
+ */
+static int max14916_iio_read_scale(void *dev, char *buf, uint32_t len,
+				   const struct iio_ch_info *channel,
+				   intptr_t priv)
+{
+	int32_t val = 1;
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+}
+
+int max14916_iio_setup_channels(struct max14916_iio_desc *desc)
+{
+	struct iio_channel *max14916_iio_channels;
+	uint32_t enabled_ch = 0;
+	uint32_t ch_offset = 0;
+	uint32_t i;
+
+	for (i = 0; i < MAX14916_CHANNELS; i++)
+		if (desc->ch_enable[i])
+			enabled_ch++;
+
+	max14916_iio_channels = no_os_calloc(enabled_ch,
+					     sizeof(*max14916_iio_channels));
+	if (!max14916_iio_channels)
+		return -ENOMEM;
+
+	for (i = 0; i < MAX14916_CHANNELS; i++) {
+		if (desc->ch_enable[i]) {
+			max14916_iio_channels[ch_offset] = (struct iio_channel)MAX14916_CHANNEL(i);
+			max14916_iio_channels[ch_offset].attributes = max14916_attrs;
+			ch_offset++;
+		}
+		max14916_iio_channels[i].ch_out = 1;
```

Comment:
you're accessing potentially unallocated memory

Reply:
moving this line in the if condition above.